File: src/designpatterns/adapter/adapter_logger.cpp
#include "json_logger.hpp"
#include "xml_logger.hpp"

class LoggerAdapter : public JSONLogger {
private:
    XMLLogger& legacyLogger;

public:
    explicit LoggerAdapter(XMLLogger& logger) : legacyLogger(logger) {}

    void logMessage(const std::string& message) override {
        legacyLogger.log(message);
    }
};

int main() {
    XMLLogger xmllogger;
    LoggerAdapter adapter(xmllogger);

    adapter.logMessage("<message>hello</message>");
    return 0;
}

File: src/designpatterns/adapter/AdapterTS/adapter_logger.ts
/* ADAPTER */
import type { JSONLogger } from './json_logger.ts';
import { XMLLogger } from './xml_logger.ts';

class LoggerAdapter implements JSONLogger {
    constructor(private legacyLogger: XMLLogger) {}

    logMessage(message: string): void {
        this.legacyLogger.log(message);
    }
    
    // Client code
    public static main(): void {
        const logger: JSONLogger = new LoggerAdapter(new XMLLogger());
        logger.logMessage('<message>hello</message>');
    }
}

LoggerAdapter.main();